New participation keys (#57)
algorandfoundation/specs: dev/partkey.md

And this verified credential is wrapped in a $\Vote$ struct with _Raw Vote_ 
($\mathrm{R}$), _Verified Credential_ ($\Credential$), and _Signature_ ($\Sig$).

## Algorand State Proof Keys 
### Algorand's Committable Ephemeral Keys Scheme - Merkle Keystore

> Algorand archives [forward security](https://en.wikipedia.org/wiki/Forward_secrecy) using the Merkle Keystore scheme. This scheme consists of designate an ephemeral key for each round in which it will be used. The scheme uses a Merkle tree to generate a commitment on those keys. 

Reviewer: ```suggestion
Algorand achieves [forward security](https://en.wikipedia.org/wiki/Forward_secrecy) using the Merkle Keystore scheme. This scheme consists in using a different ephemeral key for each round in which it will be used. The scheme uses a Merkle tree to generate a commitment on those keys. 
```

Reviewer: it says "round" here, but "participation period" later.

My understanding is that "participation period" is actually longer than rounds. Can you please define "participation period" and make everything consistent?

> The private key must be deleted be the done in order the completely archives forward secrecy.

Reviewer: ```suggestion
The private key must be deleted after it rounds passes in order the completely archives forward secrecy.
```

> The ephemeral key can be generated using one of Algorand's supported digital signature algorithms.

Reviewer: Isn't the state proof using FALCON in all cases?
Or do you allow to be used also with ED25519?
Would both proofs be generated?

In order to bound the proofs on this tree, the tree's depth is bound to 16. Hence, the maximum number of keys which will be created is 2^16.

#### Public Commitment

> The scheme generate keys that correspond to a participation period. Given _FirstValidRound_, _LastValidRound_ and an _Interval_, a key is generated for each _Round_ that holds:\newline

Reviewer: ```suggestion
The scheme generates keys that correspond to a participation period. Given _FirstValidRound_, _LastValidRound_ and an _Interval_, a key is generated for each _Round_ that holds:\newline
```

 _FirstValidRound_ $\leq$  _Round_ $\leq$ _LastValidRound_ and  _Round_  % _Interval_  = 0

> Note that there might not be a commitment for certain period (i.e does not include any round that holds _Round_  % _Interval_  = 0) and the value of 0 represents an empty commitment 

Reviewer: ```suggestion
Note that there might not be a commitment for certain period (i.e does not include any round such that _Round_  % _Interval_  = 0) and the value of 0 represents an empty commitment 
```

What is the purpose of excluding such rounds?

Author: My intention was to state that 
1 the value 0 represents an empty commitment.
2 user might get an empty commitment if he tries to generate keys for short period (i.e: a period that doesn't include a round such that: round % interval ==0.)

did that make sense?

Reviewer: I see

After generating the public key, the scheme uses those keys as leaf to create a Merkle tree in the following manner: \newline

> _leaf_$_{i}$ = hash("KP" || MsgPackEncode(_P_$_{k_{i}}$ || _Round_)) for each corresponding round.

Reviewer: What does "MsgPackEncode does?

The scheme uses the sumhash function for calculating hash on leaves and on internal nodes.

#### Signatures

A Signature in the scheme consist of the following elements: 

- _Signature_ is a signature generated by the underlying cryptographic scheme. 

>    MsgPackEncode(_signature_||_algorithm type_)

Reviewer: What is `_algorithm type_`? What are the possible values?

- _VerifyingKey_ is a public key generated by the underlying cryptographic scheme. 

>   MsgPackEncode(_algorithm type_||_public key_)

Reviewer: Why is `_algorithm type_` at the beginning there, and at the end before?

- _Proof_ is an array of hash results used as a proof for the public key within a specific round. 

>   MsgPackEncode([_digest_$_{0}$, ..., _digest_$_{n}$]||_hash type_)

Reviewer: Same questions for `hash type`?
Is proof the same as the one defined for Merkle Trees above?
